vscode-lean4/src/utils/batch.ts

```diff
@@ -39,3 +39,55 @@ export async function batchExecute(
     });
 
 }
+
+export async function testExecute(
+    toolchainPath: string,
+    args: any[],
+    workingDirectory: string,
+    delay: number = 1000): Promise<number> {
```

Comment:
If you start `lake serve` on a new lean version, then elan will certainly take more than one second to download it.  Also `lake serve` seems to do some initialization, which we probably shouldn't interrupt.  Do we ensure the download happens before trying `lake serve`?  Do we need the timeout at all?

Reply:
Yep this is the part I'm most unsure about, what I really want is to be able to ask Lake "can you run a server here" and have it answer "yes".  The timeout is needed because "lake serve" runs forever :-) I think the install should not happen because I have already called "lean --version" in the extension activate `const versionInfo = await installer.testLeanVersion();` and any required install should have happened there (with nice notification progress, etc).

Reply:
Ooo, I wonder if this can be the test that I do instead: `lake env lean --version` this one terminates and fails when there is no lakefile.lean and it fails when the lakefile.lean  contains a syntax error...

Reply:
It works, so I'm moving to that approach, much better than having to kill the "lake serve" process.

Reply:
That's not going to be better than checking for the existence of `lakefile.lean`.  What I'm worried about is versions of lake that don't have `lake serve` (it's a somewhat recent addition).

It should be enough to just close the standard input of `lake serve`.  At least locally, `lake server </dev/null` terminates promptly.

Reply:
Ok, I'll see if I can figure out how to do that in typescript...

Reply:
Rats, lake returns an error code in that case, so now I have to check for the english message 'Watchdog error: Cannot read LSP request: Stream was closed\n'.  Yuk.

Reply:
I pushed the changes.

Reply:
> That's not going to be better than checking for the existence of `lakefile.lean`. What I'm worried about is versions of lake that don't have `lake serve` (it's a somewhat recent addition).

Just to mention an alternative, we could also ask @tydeu to tag a new version for `serve` and check for that.

Reply:
> we could also ask @tydeu to tag a new version for serve and check for that.

I think that would be the best solution.  It's also more future-proof.  (Mac just needs to make sure the lake version is bumped when the `lake serve` API changes.)

Reply:
I'm not following?  I went with the idea of closing the stdin stream, which terminates `lake serve`.  The only thing I'd like to improve is make `lake` or `lean` not spit out out error when stdin is closed.  It should just quietly terminate because there's nothing more to read and return 0.

Reply:
@gebner , I'm calling this done for now, please reopen if you disagree, thanks.

Reply:
> I'm not following?

I think we should parse the output of `lake version`.  This also allows us to adapt if in the future the command-line syntax of `lake serve` changes (e.g. dropping the `--`, or allowing other arguments, etc.).

Reply:
Sorry I'm confused, are you asking me to not use `lake serve` and go back to my previous code using `lake env lean --version` ?  This is what is currently checked into this PR branch:
```typescript
 const expectedError = 'Watchdog error: Cannot read LSP request: Stream was closed\n';
            const rc = await testExecute(cmd, ['serve'], folder.fsPath, this.outputChannel, true, expectedError);
            if (rc !== 0) {
                const failover = 'Lake failed, using lean instead.'
                console.log(failover);
                if (this.outputChannel) this.outputChannel.appendLine(failover);
                useLake = false;
            }
```

Reply:
Sorry, what I meant was neither of the two.  My last suggestion was to parse the output of `lake --version`.  If it's >=4.0.0 (which Mac will hopefully release soon), then use `lake serve`, otherwise fall back to `lean --server`.

Reply:
Note that `lake env lean --version` is different from `lake --version`.

Reply:
Ok, but why? The code I have now works...

Reply:
Ok, I pushed this change.  See [this commit](https://github.com/lovettchris/vscode-lean4/commit/aaf10145197c2495ea6651e14ff210a353fde745)

Reply:
@lovettchris you should take into account the suffix for version checking `3.0.0` =/= `3.0.0-pre`. My plan is to bump the officially bump the version to `3.0.0` this weekend (which will be the first version with official `lake serve` support).

Reply:
Ok, I added parsing of the pre-release label in [Version.ts](https://github.com/leanprover/vscode-lean4/blob/735e54b18267ca87d8a8fffeb409e612b30de9f2/vscode-lean4/src/utils/version.ts) and switch the check in leanclient.ts to "3.0.0" and tested that it skips "lake serve" on my machine due to the fact that it finds version "3.0.0-pre".